cli/src/commands/redo.rs

```diff
@@ -0,0 +1,160 @@
+    //   early redo-operation restored to.
+    // - If the operation to restore to is a redo-operation itself, restore directly
+    //   to the original operation. This avoids creating a linked list of
+    //   redo-operations, which subsequently may have to be walked with an
+    //   inefficient loop.
+    //
+    // This described behavior leads to "jumping over" old redo-stacks if the
+    // current one grows into it. Consider the following op-log example, where
+    // redo-stacks are shown on the left and undo-stacks on the right:
+    //
+    // +------- "redo: restore C" * I
+    // |                          |
+    // | +----- "redo: restore D" * H
+    // | |                        |
+    // | |                        * G "undo: restore A" -------+
+    // | |                        |                            |
+    // | |   +- "redo: restore D" * F                          |
+    // | |   |                    |                            |
+    // | |   |                    * E "undo: restore A" ---+   |
+    // | |   |                    |                        |   |
+    // | +-> +---------------->   * D "undo: restore B" -+ |   |
+    // |                          |                      | |   |
+    // +---------------------->   * C                    | |   |
+    //                            |                      | |   |
+    //                            * B   <----------------+ |   |
+    //                            |                        |   |
+    //                            * A   <------------------+ <-+
+    //
+    // The first interesting operation here is I:
+    // - Attempt to redo G.
+    // - G is an undo-operation, attempt to restore its parent F.
+    // - F is a redo-opreation. Restore its original operation D, instead of F.
+    //
+    // The operation H is also noteworthy:
+    // - Attempt to redo H.
+    // - H is a redo-operation restoring to D, so attempt to redo D.
+    // - D is an undo-operation. Redo it by restoring its parent C.
+    //
+    if let Some(id_of_restored_op) = op_to_redo
+        .metadata()
+        .description
+        .strip_prefix(REDO_OP_DESC_PREFIX)
+    {
+        let Some(id_of_restored_op) = OperationId::try_from_hex(id_of_restored_op) else {
+            return Err(internal_error(
+                "Failed to parse ID of restored operation in redo-stack",
+            ));
+        };
+        op_to_redo = workspace_command
+            .repo()
+            .loader()
+            .load_operation(&id_of_restored_op)?;
+    }
+
+    if !op_to_redo
+        .metadata()
+        .description
+        .starts_with(UNDO_OP_DESC_PREFIX)
+    {
+        // cannot redo a non-undo-operation
+        return Err(user_error("Nothing to redo."));
+    }
+
+    let mut op_to_restore = match op_to_redo.parents().at_most_one().ok().flatten() {
+        Some(parent_of_op_to_redo) => parent_of_op_to_redo?,
+        None => return Err(user_error("Nothing to redo.")),
```

Comment:
nit: maybe this should be an internal error? An undo operation should have a single parent.